### DISCUSSION TITLE
Traverse a dynamic json tree and collect all the objects with keys of certain value

### DISCUSSION THREAD

**Discussion by KrushnaOrigoNetworks** [schibsted/jslt#215] [Q&A]
I would like to collect all objects with keys "meta" in this case

"obj": {
    "properties": {
        "x": {
            "properties": {
                "x1": {
                    "values": [
                        {
                            "name": "My name",
                            "meta": {
                                "id": "Id15",
                                "time": "2021-08-11T19:38:55.953Z"
                            },
                            "value": "Justin"
                        }
                    ],
                    "name": "Nname",
                    "description": "Description",
                    "meta": {
                        "id": "Id12",
                        "time": "2021-08-11T19:38:55.953Z"
                    }
                },
                "x2": {
                    "values": [
                        {
                            "name": "Groovy",
                            "description": "Groovy description",
                            "meta": {
                                "id": "Id17",
                                "time": "2021-08-11T19:38:55.953Z"
                            },
                            "value": "groovy"
                        },
                        {
                            "name": "Java",
                            "meta": {
                                "id": "Id16",
                                "time": "2021-08-11T19:38:55.953Z"
                            },
                            "value": "java"
                        }
                    ],
                    "name": "Skills",
                    "description": "Skills description",
                    "meta": {
                        "id": "Id10",
                        "time": "2021-08-11T19:38:55.953Z"
                    }
                },
                "x3": {
                    "name": "FirstName",
                    "age": 25,
                    "meta": {
                        "id": "Id14",
                        "time": "2021-08-11T19:38:55.953Z"
                    }
                },
                "id": {
                    "name": "ID",
                    "meta": {
                        "id": "Id11",
                        "time": "2021-08-11T19:38:55.953Z"
                    }
                }
            }
        }
    }
}

**Comment by larsga** [comment]
Something like this?

```
{for (.obj.properties.x.properties)
  .key : .value
  if (.value.meta)
}
```

**Reply by KrushnaOrigoNetworks** [reply]
Thanks, Larsga.
That was just an example. The tree is created dynamically and so need to traverse the tree looking for the key "meta".
Greatly appreciate your help and prompt answer.

**Reply by larsga** [reply]
For the traversal part you can look at [this discussion](https://github.com/schibsted/jslt/discussions/213).

**Reply by KrushnaOrigoNetworks** [reply]
Thank you, Larsga.
Do you have any suggestions on how to collect the data as we traverse?

**Reply by larsga** [reply]
The easiest thing is to simply make the function return each object as you want it to be in the output. Basically:

```
def traverse(node)
  if (is-object($node))
    // do meta cleaning here
  else
    $node

traverse(.)
```

**Reply by KrushnaOrigoNetworks** [reply]
Thank you, Larsga. 
I implemented this way. It works with some empty objects that I can ignore.

def traverse(obj)
    if (is-object($obj))
        [{for ($obj) 
           .key: .value if (.key == "meta")}] +
         [for ($obj) 
           traverse(.value) if (.key != "meta")]
   else if (is-array($obj))
        [for ($obj) 
            traverse(.)
        ]
   else 
       []
                     
flatten(traverse(.))
